src/java/org/apache/cassandra/service/accord/AccordTopologyService.java

```diff
@@ -0,0 +1,204 @@
+
+    private SortedArrayList<Node.Id> previouslyRemovedIds = SortedArrayList.ofSorted();
+
+    private enum State { INITIALIZED, STARTED, SHUTDOWN }
+
+    @GuardedBy("this")
+    private State state = State.INITIALIZED;
+
+    public AccordTopologyService(Node.Id node, AccordEndpointMapper endpointMapper, MessageDelivery messagingService, ScheduledExecutorPlus scheduledTasks)
+    {
+        this.syncPropagator = new AccordSyncPropagator(node, endpointMapper, messagingService, scheduledTasks);
+        this.watermarkCollector = new WatermarkCollector();
+    }
+
+    public AccordTopologyService(Node.Id node, AccordEndpointMapper endpointMapper)
+    {
+        this(node, endpointMapper, MessagingService.instance(), ScheduledExecutors.scheduledTasks);
+    }
+
+    /**
+     * On restart, loads topologies. On bootstrap, discovers existing topologies and initializes the node.
+     */
+    public void onStartup(Node node)
+    {
+        SortedArrayList<Node.Id> removed = node.topology().current().removedIds();
+        synchronized (this)
+        {
+            Invariants.require(state == State.INITIALIZED, "Expected state to be INITIALIZED but was %s", state);
+            state = State.STARTED;
+            previouslyRemovedIds = removed;
+            node.topology().addListener(watermarkCollector);
+            node.topology().addListener(syncPropagator);
+        }
+        syncPropagator.onNodesRemoved(removed);
+    }
+
+    @Override
+    public synchronized boolean isTerminated()
+    {
+        return state == State.SHUTDOWN;
+    }
+
+    @Override
+    public synchronized void shutdown()
+    {
+        if (isTerminated())
+            return;
+        state = State.SHUTDOWN;
+    }
+
+    @Override
+    public Object shutdownNow()
+    {
+        shutdown();
+        return null;
+    }
+
+    @Override
+    public boolean awaitTermination(long timeout, TimeUnit units) throws InterruptedException
+    {
+        return isTerminated();
+    }
+
+    @Override
+    public void onReceived(Topology topology)
+    {
+        SortedArrayList<Node.Id> newlyRemoved;
+        synchronized (this)
+        {
+            newlyRemoved = topology.removedIds().without(previouslyRemovedIds);
+            previouslyRemovedIds = topology.removedIds().with(previouslyRemovedIds);
+        }
+        syncPropagator.onNodesRemoved(newlyRemoved);
+    }
+
+    @Override
+    public AsyncResult<Topology> fetchTopologyForEpoch(long epoch)
+    {
+        SettableByCallback<Topology> result = new SettableByCallback<>();
+        fetchTopologyAsync(epoch, result);
+        return result;
+    }
+
+    private void fetchTopologyAsync(long epoch, BiConsumer<? super Topology, ? super Throwable> onResult)
+    {
+        // It's not safe for this to block on CMS so for now pick a thread pool to handle it
+        Stage.ACCORD_MIGRATION.execute(() -> {
+            try
+            {
+                if (ClusterMetadata.current().epoch.getEpoch() < epoch)
```

Comment:
nit: can move `metadata` variable declaration from L167 and put it before `try` to save a call, also useful for clarity that we always have just one epoch here.